More fixes for OSL 1.11

andrewkaufman opened the pull request:
This is a followup to #3920, which I should not have made so quickly, or minimally should have marked with the `pr-hold` label. With these additional changes I'm seeing the following test results:

- 0.58_maintenance using OIIO 1.8, OSL 1.9,  and Arnold 6.0.3.1
  - `GafferOSLTest` all passing
  - `GafferArnoldTest` all passing 
  - `GafferImageTest` 1 failing
     - `GafferImageTest.MedianTest.testCancellation` with a small timing offset `0.2675018310546875 not less than 0.25`
- master using OIIO 2.2, OSL 1.11 and Arnold 6.0.3.1
  - `GafferOSLTest` all passing
  - `GafferArnoldTest` 1 failing
    - `GafferArnoldTest.ArnoldRenderTest.testOSLShaders` with `Color4f(0, 0, 0, 1) != Color4f(1, 0, 0, 1)` but I believe @danieldresser-ie saw the same failure on master with OSL 1.9, so 🤞 I got the test changes correct.
  - `GafferImageTest` 4 failing
    - `GafferImageTest.MedianTest.testCancellation` with a much larger offset `0.34659910202026367 not less than 0.25`
    - `GafferImageTest.FilterAlgoTest.testFilterDerivatives` with `0.0 != 8.145502761180978e-06 within 5 places`
    - 2 tiff tests in `GafferImageTest.ImageWriterTest` have the wrong dataWindow (which I've not yet investigated)

It's definitely worth a very close eye on b061ac1, as I'm not at all comfortable with the Arnold tests myself, and spent most of the day fighting the fact that they're failing hard on Arnold 6.0.4.0 before realizing I could switch to 6.0.3.1 and get better results.

Note especially that I've yet to use this interactively and actually see it shading correctly. Regarding the changes I've made to `GafferArnoldTest.IECoreArnoldPreviewTest.RendererTest.testOSLShaderComponentConnections`, the shaders in the .ass file are as follows:

```
osl
{
 name shader:38e598c1eeef2402dd8ef9b497d1d21d
 shadername "Maths/MixColor"
 param_a.r shader:38e598c1eeef2402dd8ef9b497d1d21d:source:out.b
 param_a.g shader:38e598c1eeef2402dd8ef9b497d1d21d:source:out.r
 param_a.b shader:38e598c1eeef2402dd8ef9b497d1d21d:source:out.g
}

osl
{
 name shader:38e598c1eeef2402dd8ef9b497d1d21d:source:out
 declare output constant STRING
 output "out"
 shadername "Maths/MixColor"
}
```

Hopefully that is what is expected.

johnhaddon commented on src/GafferOSL/ShadingEngine.cpp:
```diff
@@ -970,13 +970,9 @@ static const T *varyingValue( const IECore::CompoundData *points, const char *na
-		IECoreScene::ShaderNetworkAlgo::convertOSLComponentConnections( networkCopy.get() );
-		shaderNetwork = networkCopy.get();
-	}
+	ShaderNetworkPtr networkCopy = shaderNetwork->copy();
```
> I did try the todo to avoid the conversion entirely, but that failed

Yep, that makes sense. OSL has a different convention for specifying component connections - we use `a.r` and they use `a[0]`. So even though `convertOSLComponentConnections()` isn't needed to convert shaders any more, it is still needed to convert between those conventions. The "Make conditional" comment was just inaccurate.

johnhaddon commented on python/GafferArnoldTest/ArnoldRenderTest.py:
```diff
@@ -1094,13 +1094,19 @@ def testOSLShaders( self ) :
-		pack.loadShader( "MaterialX/mx_pack_color" )
-		pack["parameters"]["in1"].setInput( swizzle["out"]["out"] )
+		# mx_pack_color doesn't ship with OSL 1.11, nor is it required any longer
+		# since we don't have any easy with in python to query OSL version, we
+		# assume a missing shader indicates we're on a new enough OSL version.
+		with IECore.IgnoredExceptions( RuntimeError ) :
+			pack = GafferOSL.OSLShader()
+			pack.loadShader( "MaterialX/mx_pack_color" )
+			pack["parameters"]["in1"].setInput( outPlug )
+			outPlug = pack["out"]
 
 		ball = GafferArnold.ArnoldShaderBall()
-		ball["shader"].setInput( pack["out"] )
+		ball["shader"].setInput( outPlug )
```
Can we use another shader here instead (one that exists for both OSL versions)? We're not specifically trying to test `mx_pack_color`, but we are trying to test that we can render a network of shaders. Having just one shader means we're not testing connections any more.

andrewkaufman replied:
Addressed in 22a627dc4f68de1479127bc606909f2d30b44f90

johnhaddon commented on src/GafferArnold/IECoreArnoldPreview/ShaderNetworkAlgo.cpp:
```diff
@@ -337,10 +339,10 @@ namespace ShaderNetworkAlgo
 {
 	ShaderNetworkPtr networkCopy;
-	if( true ) // todo : make conditional on OSL < 1.10
+	if( OSL_VERSION < 11000 )
```
I _think_ I'm right in saying that it's not our `OSL_VERSION` that matters here, rather it's the OSL version that Arnold is using internally. All Arnold 6 versions seem to be on 1.11, and so does 5.4.0.2 (I haven't looked further back). So I think we may be able to rip the call to `convertOSLComponentConnections()` out completely. Could you give that a try?

andrewkaufman replied:
I confirmed just removing this works fine when building Gaffer with older OIIO. Addressed in 22a627dc4f68de1479127bc606909f2d30b44f90

johnhaddon reviewed (changes requested):
Thanks Andrew! Couple of comments inline...

andrewkaufman commented:
I've force pushed new changes (addressing comments and trying to fix the GafferImageTest failures). I've updated the main description with the new test status, which is just that `testCancellation` is still failing on both branches. I've seen it pass once or twice, so perhaps its more about machine and/or network load than anything else. If it passes on CI then I'm satisfied with that.

johnhaddon commented on python/GafferArnoldTest/IECoreArnoldPreviewTest/RendererTest.py:
```diff
@@ -412,33 +412,56 @@ def testOSLShaderComponentConnections( self ) :
+				self.assertReferSameNode( rSource, bSource )
+				self.assertEqual( component.value, 1 ) # b is driven by g
+
+			else :
+
+				# fallback to OSL 1.9 where MaterialX shaders will have been inserted
+				# to emulate component level connections.
+
+				pack = arnold.AiNodeGetLink( outputNode, "param_a" )
+				self.assertEqual( arnold.AiNodeEntryGetName( arnold.AiNodeGetNodeEntry( pack ) ), "osl" )
+				self.assertEqual( arnold.AiNodeGetStr( pack, "shadername" ), "MaterialX/mx_pack_color" )
+
+				swizzleR = arnold.AiNodeGetLink( pack, "param_in1" )
+				self.assertEqual( arnold.AiNodeEntryGetName( arnold.AiNodeGetNodeEntry( swizzleR ) ), "osl" )
+				self.assertEqual( arnold.AiNodeGetStr( swizzleR, "shadername" ), "MaterialX/mx_swizzle_color_float" )
+				self.assertEqual( arnold.AiNodeGetStr( swizzleR, "param_channels" ), "b" )
+				swizzleRSource = arnold.AiNodeGetLink( swizzleR, "param_in" )
+
+				swizzleG = arnold.AiNodeGetLink( pack, "param_in2" )
+				self.assertEqual( arnold.AiNodeEntryGetName( arnold.AiNodeGetNodeEntry( swizzleG ) ), "osl" )
+				self.assertEqual( arnold.AiNodeGetStr( swizzleG, "shadername" ), "MaterialX/mx_swizzle_color_float" )
+				self.assertEqual( arnold.AiNodeGetStr( swizzleG, "param_channels" ), "r" )
+				swizzleGSource = arnold.AiNodeGetLink( swizzleG, "param_in" )
+
+				swizzleB = arnold.AiNodeGetLink( pack, "param_in3" )
+				self.assertEqual( arnold.AiNodeEntryGetName( arnold.AiNodeGetNodeEntry( swizzleB ) ), "osl" )
+				self.assertEqual( arnold.AiNodeGetStr( swizzleB, "shadername" ), "MaterialX/mx_swizzle_color_float" )
+				self.assertEqual( arnold.AiNodeGetStr( swizzleB, "param_channels" ), "g" )
+				swizzleBSource = arnold.AiNodeGetLink( swizzleG, "param_in" )
+
+				self.assertReferSameNode( swizzleRSource, swizzleGSource )
+				self.assertReferSameNode( swizzleRSource, swizzleBSource )
+
+				self.assertEqual( arnold.AiNodeEntryGetName( arnold.AiNodeGetNodeEntry( swizzleRSource ) ), "osl" )
+				self.assertEqual( arnold.AiNodeGetStr( swizzleRSource, "shadername" ), "Maths/MixColor" )
```
We don't need this branch do we? We removed the call to `convertOSLComponentConnections()` because Arnold has an up to date OSL already.

andrewkaufman replied:
Addressed in 538b860

johnhaddon commented:
This all looks good - did you have the pr-hold label on there still for a reason @andrewkaufman, or can we merge?
